File: leetcode/design/379_Design_Phone_Directory/codes/queue.py
from collections import deque

class PhoneDirectory:
    def __init__(self, maxNumbers: int):
        self.isSlotAvailable = [True] * maxNumbers
        self.slotsAvailableQueue = deque(range(maxNumbers))

    def get(self) -> int:
        if not self.slotsAvailableQueue:
            return -1
        slot = self.slotsAvailableQueue.popleft()
        self.isSlotAvailable[slot] = False
        return slot

    def check(self, number: int) -> bool:
        return self.isSlotAvailable[number]

    def release(self, number: int) -> None:
        if self.isSlotAvailable[number]:
            return
        self.slotsAvailableQueue.append(number)
        self.isSlotAvailable[number] = True

# Your PhoneDirectory object will be instantiated and called as such:
# obj = PhoneDirectory(maxNumbers)
# param_1 = obj.get()
# param_2 = obj.check(number)
# obj.release(number)

File: leetcode/design/379_Design_Phone_Directory/codes/queue.cpp
#include <vector>
#include <queue>
using namespace std;

class PhoneDirectory {
    queue<int> slotsAvailableQueue;
    vector<bool> isSlotAvailable;

public:
    PhoneDirectory(int maxNumbers) {
        isSlotAvailable = vector<bool>(maxNumbers, true);
        for (int i = 0; i < maxNumbers; ++i) {
            slotsAvailableQueue.push(i);
        }
    }

    int get() {
        if (slotsAvailableQueue.empty()) {
            return -1;
        }
        int slot = slotsAvailableQueue.front();
        slotsAvailableQueue.pop();
        isSlotAvailable[slot] = false;
        return slot;
    }

    bool check(int number) {
        return isSlotAvailable[number];
    }

    void release(int number) {
        if (isSlotAvailable[number]) {
            return;
        }
        slotsAvailableQueue.push(number);
        isSlotAvailable[number] = true;
    }
};

/**
 * Your PhoneDirectory object will be instantiated and called as such:
 * PhoneDirectory* obj = new PhoneDirectory(maxNumbers);
 * int param_1 = obj->get();
 * bool param_2 = obj->check(number);
 * obj->release(number);
 */